cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,420 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
 }
 
 /**
- * Same as RandomTraverser except that the number of steps taken is bounded by {@code maxSteps}
+ * Same as RandomTraverser except that the number of steps taken is bounded by `maxSteps`
  */
-class RandomBoundedTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int], val maxSteps: Long,
+class RandomBoundedTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int], maxStepss: Long,
     walkParams: RandomWalkParams) extends RandomTraverser(graph, dir, homeNodeIds,
     walkParams.resetProbability, walkParams.maxNumEdgesThresh, walkParams.visitSameNodeOnce,
     walkParams.randNumGen, walkParams.maxDepth,
-    walkParams.filterHomeNodeByNumEdges) with BoundedIterator[Node]
+    walkParams.filterHomeNodeByNumEdges) with BoundedIterator[Node] {
+  override def maxSteps: Option[Long] = Some(maxStepss)
+}
 
 /**
- * A traverser that keeps track of distance from homeNodeIds. If {@code onlyOnce} is true,
- * the same node is not traversed again.
+ * Optional limit used for limiting depth or degree of nodes that are added
+ * to the queue.
  */
-abstract class DistanceTraverser[T](graph: Graph, homeNodeIds: Seq[Int], onlyOnce: Boolean)
-    extends Traverser { self =>
+class Limit(limit: Option[Int]) {
+  def isLimitReached(value: Int): Boolean = limit match {
+    case Some(d) if value >= d => true
+    case _ => false
+  }
+}
 
-  private val distanceTracker = new IntInfoKeeper(self.onlyOnce)
+/**
+ * We use graph coloring in the QueueBasedTraverser. We use
+ * 3 colors defined as `NodeColor`.
+ */
+object NodeColor extends Enumeration {
+  type Color =  Value
+  val Unenqueued = Value
+  val Enqueued = Value
+  val Visited = Value
+}
 
-  protected def init(defaultInfo: T) {
-    homeNodeIds foreach { id => visitPotentialNode(id, defaultInfo, 0) }
-  }
+/**
+ * Stores information of three colors assigned to node ids (Ints).
+ *
+ * Initially every node has color `Unseen`.
```

Comment:
I would stay with Unenqueued. If we have Unseen, Enqueued it can confusing whether a node can be Seen and still not enqueued. In all our version we enqueue a node when we see it for the first time. Is this ok?